CIScripts/Common/Invoke-NativeCommand.ps1

```diff
@@ -18,30 +21,84 @@ function Invoke-NativeCommand {
+        )
+
+        if ($Session) {
+            Invoke-Command -Session $Session -ScriptBlock $ScriptBlock
+        } else {
+            Invoke-Command -ScriptBlock $ScriptBlock
+        }
+    }
 
-    # If an executable in $ScriptBlock wouldn't be found then while checking $LastExitCode
-    # we would be checking the exit code of a previous command. To avoid this we clear $LastExitCode.
-    $Global:LastExitCode = $null
+    Invoke-CommandLocalOrRemote -Session $Session -ScriptBlock {
+        # If an executable in $ScriptBlock wouldn't be found then while checking $LastExitCode
+        # we would be checking the exit code of a previous command. To avoid this we clear $LastExitCode.
+        $Global:LastExitCode = $null
+    }
+
+    # We need to backup only the ErrorActionPreference only in remote case,
+    # as all changes of this variable locally are scoped to a function,
+    # but when using Invoke-Comand -Session $Session, their scope is
+    # for a whole lifetime of a session.
+    if ($Session) {
+        $OldRemoteErrorAction = Invoke-Command -Session $Session { $ErrorActionPreference }
+    }
 
-    & {
-        # Since we're redirecting stderr to stdout we shouldn't have to set ErrorActionPreference
-        # but because of a bug in Powershell we have to.
-        # https://github.com/PowerShell/PowerShell/issues/4002
-        $ErrorActionPreference = "Continue"
+    # Since we're redirecting stderr to stdout we shouldn't have to set ErrorActionPreference
+    # but because of a bug in Powershell we have to.
+    # https://github.com/PowerShell/PowerShell/issues/4002
+    if ($Session) {
+        Invoke-Command -Session $Session { $ErrorActionPreference = "Continue" }
+    }
+    $ErrorActionPreference = "Continue"
 
+    try {
         # We redirect stderr to stdout so nothing is added to $Error.
         # We do this to be compliant to durable-task-plugin 1.18.
-        & $ScriptBlock 2>&1
+        if ($CaptureOutput) {
+            $Output = @()
+            $Output += Invoke-CommandLocalOrRemote -Session $Session -ScriptBlock $ScriptBlock 2>&1
+        } else {
+            Invoke-CommandLocalOrRemote -Session $Session -ScriptBlock $ScriptBlock 2>&1 | Write-Host
+        }
+    }
+    finally {
+        if ($Session) {
+            Invoke-Command -Session $Session { $ErrorActionPreference = $Using:OldRemoteErrorAction }
+        }
+        $ErrorActionPreference = "Stop"
+    }
+
+    $ExitCode = Invoke-CommandLocalOrRemote -Session $Session {
+        $ExitCode = $Global:LastExitCode
+        # We clear it to be compliant with durable-task-plugin up to 1.17
+        $Global:LastExitCode = $null
+        $ExitCode
     }
 
-    if ($AllowNonZero -eq $false -and $LastExitCode -ne 0) {
-        throw "Command ``$ScriptBlock`` failed with exitcode: $LastExitCode"
+    if ($AllowNonZero -eq $false -and $ExitCode -ne 0) {
+        throw "Command ``$ScriptBlock`` failed with exitcode: $ExitCode"
     }
 
+    $ReturnDict = @{}
```

Comment:
[Change] I wonder if we could just ```return @{ExitCode=..., Output =....}``` if CaptureOutput is true. When CaptureOutput is false, we can just return nothing. It may simplify a code a bit. What do you think about that ?

Reply:
I thought about that before, here's my reasoning why always returning a full dict is not a perfect solution:
1. If we always return a full dict, the simple
   ```powershell
   Invoke-NativeCommand { foo.exe }
   ```
   would always print something like that to output/logs:
   ```
   Name                           Value
   ----                           -----
   ExitCode                       0
   Output                         {}
   ```
   And that would require adding `| Out-Null`, which in my opinion, would be just confusing.
2. The `Output` field present (but empty) when somebody just forgot to capture output would result in thinking "why this process hasn't produce any output, I'm sure it had!".
   If the `Output` field is just present, as now, the user would get an error about missing field, which is a nicer experience, imho.

Reply:
Sorry, I've misread your suggestion.

Reply:
What about `-AllowNonZero` without `-CaptureOutput`?

Reply:
In my opinion since -CaputreOutput is not enabled, than caller is not interested in exit code, so I would return nothing